### PR TITLE
Big rewrite and consul integration

### PR DESCRIPTION
**Pull request by OisinA**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM golang:latest
```
Comment (Strum355) [inline Dockerfile]: pls pin go version :)

File: main.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+)
+
+func main() {
+	// Initialise logger
+	log.InitSimpleLogger(&log.Config{})
```
Comment (Strum355) [inline main.go]: pls have this use JSON logger if in production :) recommend spf13/viper for loading and using config. see windlass codebase for example

File: main.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+	// Gather environment variables
+	port := "8080"
```
Comment (Strum355) [inline main.go]: this port can be an spf13/viper setting :)

**Review by Strum355 — CHANGES_REQUESTED** [review]

File: api/jail.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	jail := chi.URLParam(r, "jail")
+
+	currentFailed, totalFailed, fileList, currentBanned, totalBanned, ipList, err := a.Fail2Conn.JailStatus(jail)
+	if err != nil {
+		log.WithError(err).Error("Could not get jail")
+		a.handleError(w, r, err)
+		return
+	}
```
Comment (Strum355) [inline api/jail.go]: Do we wanna use errHandle here?

File: config/defaults.go [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+)
+
+func loadDefaults() {
+	viper.SetDefault("fail2rest.production", false)
+	viper.SetDefault("fail2rest.secret", "abcd1234")
+	viper.SetDefault("fail2rest.port", 8080)
+	viper.SetDefault("fail2rest.fail2ban", "/var/run/fail2ban/fail2ban.sock")
+	viper.SetDefault("fail2rest.consul_host", "127.0.0.1:8500")
+	viper.SetDefault("fail2rest.consul_token", "")
+	viper.SetDefault("fail2rest.token-location", "fail2rest-token")
```
Comment (Strum355) [inline config/defaults.go]: https://github.com/UCCNetworkingSociety/Windlass-worker/blob/master/app/config/defaults.go#L11-L30
See here for sample config style.

Comment (Strum355) [reply config/defaults.go]: `viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))` dont forget this to translate `ENV_LIKE_THIS` to `env.like.this`

**Review by Strum355 — CHANGES_REQUESTED** [review]
:)

File: api/api.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	"net/http"
+)
+
+type API struct {
+	Fail2Conn *fail2go.Conn
+	Secret    string
+}
+
+func (a *API) Register(r chi.Router) {
+	r.Use(a.authMiddleware)
```
Comment (Strum355) [inline api/api.go]: im wondering do we want auth for the ping endpoint?

File: main.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		os.Exit(1)
+	}
+	err = consul.Register()
+	if err != nil {
+		log.WithError(err).Error("Could not register with Consul service")
+		os.Exit(1)
+	}
+
+	// Initialise API
+	api := api.API{Fail2Conn: conn, Secret: consul.Secret}
```
Comment (Strum355) [inline main.go]: do we need to pass the secret down if its in the viper settings?
Related to point down below `v`

File: services/consul.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+		kv, _, err := c.client.KV().Get(path, &api.QueryOptions{})
+		if err != nil {
+			return err
+		}
+
+		if kv == nil {
+			return errors.New(fmt.Sprintf("key %s not set", path))
+		}
+
+		c.Secret = string(kv.Value)
```
Comment (Strum355) [inline services/consul.go]: id say assign the secret to `viper.Set("fail2rest.secret")` *or* do what youre doing here and delete the config default

**Review by Strum355 — CHANGES_REQUESTED** [review]

**Review by ArthanJans — CHANGES_REQUESTED** [review]
You forgot to add an unban endpoint

**Review by Strum355 — APPROVED** [review]

**Pull request merged**